Firestore/Swift/Tests/Codable/CodableDocumentTests.swift

```diff
@@ -0,0 +1,405 @@
+/*
+ * Copyright 2019 Google
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import Foundation
+import FirebaseFirestore
+@testable import FirebaseFirestoreSwift
+import XCTest
+
+class CodableDocumentTests: XCTestCase {
+  func roundTrip<X>(input: X, expected: [String: Any]? = nil) -> X where X: Codable {
+    var encoded = [String: Any]()
+    do {
+      encoded = try Firestore.encode(input)
+      if let expected = expected {
+        XCTAssertEqual(encoded as NSDictionary, expected as NSDictionary)
+      }
+    } catch {
+      XCTFail("Failed to encode \(X.self): error: \(error)")
+    }
+    do {
+      let decoded = try Firestore.decode(X.self, from: encoded)
+      return decoded
+    } catch {
+      XCTFail("Failed to decode \(X.self): \(error)")
+    }
+    return input // After failure
+  }
+
+  func testInt() {
+    struct Model: Codable {
+      let x: Int
+    }
+    let model = Model(x: 42)
+    let dict = ["x": 42]
+    XCTAssertEqual(model.x, roundTrip(input: model, expected: dict).x)
+  }
+
+  func testEmpty() {
+    struct Model: Codable {}
+    let model = Model()
+    let dict = [String: Any]() as NSDictionary
+    let encoded = try! Firestore.encode(model) as NSDictionary
+    XCTAssertEqual(encoded, dict)
+  }
+
+  func testNil() {
+    struct Model: Codable {
+      let x: Int?
+    }
+    let model = Model(x: nil)
+    let dict = ["x": nil] as [String: Int?]
+    let encodedDict = try! Firestore.encode(model)
+    XCTAssertNil(encodedDict["x"])
```

Comment:
So I just checked the Android SDK and I think this assertion is ambiguous. It doesn't distinguish between an "x" that's present and has a value of nil or an "x" that's absent. We really care that it's the former not the latter.

On Android, given a class like this:

```java
  private static class StringBean {
    private String value;

    public String getValue() {
      return value;
    }
  }
```

This test holds true:

```java
  @Test
  public void primitiveSerializeString() {
    StringBean bean = new StringBean();
    assertJson("{'value': null}", serialize(bean));
  }
```

This is slightly different from Swift, because nullability is implicit and default for Object-types in Java. Nevertheless it's important to preserve that the key exists with a null value in the output because null has an explicit representation in Firestore and you can't query for missing fields.

This means that we have to be very careful:
```
XCTAssertNotEqualObjects([:], ["x": nil]);
assertRoundTrip(model: model, encoded: dict)
assertDecodes([:], to: model)
```

Reply:
Handling nil properly looks challenging and is not currently supported by the JSON encoder.  See  https://stackoverflow.com/questions/47266862/encode-nil-value-as-null-with-jsonencoder and https://bugs.swift.org/browse/SR-9232. 

Presumably we don't handle this now in Objective C because NSDictionary's don't allow nil values. 

Any suggestions other than deferring `nil` support?